Support extern context impl

sagudev opened the pull request:
**Connections**
Based on https://github.com/gfx-rs/wgpu/pull/6619, fixes https://github.com/gfx-rs/wgpu/issues/6330

**Description**
This PR enables users of wgpu to provide custom Instance impl (and other wgpu types as all other types are derived from Instance->Adapter->Device), by adding `Custom` variant and custom backend that has dyn wrapper for all types: `struct DynDevice(Arc<dyn DeviceInterface>)`. I'm intermixing Custom with Dyn because I cannot really decide on name (`CustomDevice` is weird, but we already have Dyn* in wgpu-hal).
There are few rough sports:
- `RenderBundleEncoderInterface::finish` (other backends hacks around by not using dyn dispatch, but we can't: https://github.com/gfx-rs/wgpu/blob/b876b8c2816441bd53f4d787739db929f9da4dd4/wgpu/src/api/render_bundle_encoder.rs#L54-L60

**Testing**
There is wgpu-custom crate that showcases how one can create Custom wgpu types.

<!--
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're no bothering us!
-->

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [x] Run `cargo xtask test` to run tests.
- [ ] Add change to `CHANGELOG.md`. See simple instructions inside file.

sagudev commented:
TBH, I have no idea what's wrong with WASM.

sagudev commented on wgpu/src/api/instance.rs:
```diff
@@ -201,6 +203,23 @@ impl Instance {
+        Self {
+            inner: dispatch::DispatchInstance::Custom(backend::custom::DynContext::new(instance)),
+        }
+    }
```
This is triggering visibility problems for WASM, but I do not know why.

cwfitzgerald commented:
Ping me if this needs looking at or you need help before it is out of draft, going to un-assign myself for bookeeeping

sagudev commented:
In https://github.com/gfx-rs/wgpu/pull/6658/commits/79ecded9fa0f5a20fb470e6a9c27caf0fa284a63 I removed InterfaceTypes trait as it's not really needed anymore (this makes writing Custom/Dyn wrappers easier as they do not need to impl traits directly, avoiding a lot of boilerplate), all bounds are actually enforced in `dispatch_types_inner` anyway. The only useful application was making writing macros easier, but I replaced this with `mod interface_types` that has aliases for interface types (there is a lot of this pattern in servo as it was used before associate trait types), but we could also just write more stuff in macro and remove the completely. Does approach like this sounds reasonable @cwfitzgerald? I would be willing to split this into separate PR if it's ok.

cwfitzgerald commented:
Hmm... I don't love it as it makes things less structured, but can't see a strong argument against removing the trait. Will fully consider it when I review this. You can leave it as part of this PR for now.

sagudev commented:
I have alternative approach in https://github.com/sagudev/wgpu/commit/ce582808f273a18d9816d5c6e455ff0975e8b02c which keeps the trait but instead of using interfaces in bounds it uses `Dispatch<Target = dyn Interface>` (clone of Deref trait, I avoid Deref to avoid compiler magic as it makes bounds less explicit esp between Dispatch and DispatchMut). One thing to consider here is that implementations of custom context will need to provide all types, even those that are not needed (ex. if one does not need/use compute pass). Should I switch to this approach @cwfitzgerald?

cwfitzgerald commented:
> I have alternative approach in https://github.com/sagudev/wgpu/commit/ce582808f273a18d9816d5c6e455ff0975e8b02c which keeps the trait but instead of using interfaces in bounds it uses Dispatch<Target = dyn Interface> (clone of Deref trait, I avoid Deref to avoid compiler magic as it makes bounds less explicit esp between Dispatch and DispatchMut).

I'm going to leave this up to your own judgement which is cleaner and more parsable, my initial impression is that the mod { type } approach is less infrastructure and less confusing than adding the Dispatch one

sagudev commented:
@cwfitzgerald I think this is ready for concept review, it still needs more docs and maybe some polish 💅🏾.

cwfitzgerald commented:
Nice! I'm going to take this out of draft for bookkeeping

cwfitzgerald commented on wgpu-custom/Cargo.toml:
```diff
@@ -0,0 +1,17 @@
+[package]
+name = "wgpu-custom"
```
What is the purpose of this crate?

sagudev replied:
Example how to use custom context.

cwfitzgerald replied:
Can this be either an example or a test? We don't really want more crates unless there's a strong motivating reason.

sagudev replied:
This should actually be example crate like those in [examples/standalone](https://github.com/gfx-rs/wgpu/tree/trunk/examples/standalone), but at the time of writing this PR, that didn't exists.

sagudev replied:
Done in https://github.com/gfx-rs/wgpu/pull/6658/commits/52d897ea6a84165f66a217675abbc9586fd66b48, I also write a description for crate.

cwfitzgerald commented on wgpu/src/dispatch.rs:
```diff
@@ -795,55 +782,143 @@ macro_rules! dispatch_types_inner {
-                $type;
-            }
-        )*
+dispatch_types_inner! {
+    {
+        ref type DispatchInstance = Instance:InstanceInterface
     };
 }
```
can this syntax be made more terse, now that there's only ever one being done per macro invocation?

sagudev replied:
Yes we can! Done in https://github.com/gfx-rs/wgpu/pull/6658/commits/bd92b353099ed393405ce7a65cec0e0a7085129a

cwfitzgerald reviewed (changes requested):
Don't love the whole new `type` alias thing, but don't really hate it enough to say no

sagudev commented:
> Don't love the whole new `type` alias thing, but don't really hate it enough to say no

I think we could actually inline all types uses into macro invocation (all 3 of them). Would that be better?

cwfitzgerald commented:
> I think we could actually inline all types uses into macro invocation 

Hmm, yeah I think that's fine - the whole appeal to me is to make sure things have uniform naming conventions and are all filled out, so having them all in one place seems to do that trick fine.

sagudev commented:
> > I think we could actually inline all types uses into macro invocation
> 
> Hmm, yeah I think that's fine - the whole appeal to me is to make sure things have uniform naming conventions and are all filled out, so having them all in one place seems to do that trick fine.

I am not sure I understand, should I inline all types uses into macro or keep the `mod`s?

cwfitzgerald commented:
Oh sorry was unclear, inline into the macro.

cwfitzgerald commented on examples/custom-backend/Cargo.toml:
This should be either be in `standalone` or in `features`, probably `features` - I don't think this needs it needs to be on the top level. Either way it needs to be added to the `example` README

cwfitzgerald commented on wgpu/src/backend/custom.rs:
```diff
@@ -0,0 +1,88 @@
+    // cloning of arc forbidden
+    (pub mut struct $name:ident(dyn $interface:tt)) => {
+        #[derive(Debug)]
+        pub(crate) struct $name(Arc<dyn $interface>);
```
Any reason this couldn't be `Box`? If so there should be a comment

sagudev replied:
It's to obtain `Eq`,`Hash`,`Cmp` implementations. We cannot use T implementations because traits are not object safe. Will write in comment.

cwfitzgerald reviewed (changes requested):
Some small things, but I think this is good to go after that!

sagudev committed:
Support extern context impl

Signed-off-by: sagudev <16504129+sagudev@users.noreply.github.com>

cwfitzgerald reviewed (approved):
I have some futzing I want to do to the examples, but I'll do that post land.

Thanks!

DJMcNab commented:
Has this been released? I'm trying to check the changelog for when this got released

sagudev commented:
> Has this been released? I'm trying to check the changelog for when this got released

Yes it was, but there are still some rough edges: https://github.com/gfx-rs/wgpu/issues/7533
